Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch

```diff
@@ -0,0 +1,56 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: ysl3000 <yannicklamprecht@live.de>
+Date: Mon, 6 Jul 2020 22:18:04 +0200
+Subject: [PATCH] Create-HoverEvent-from-ItemStack-Entity
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index d8103ab0d542ced4f30b88ceb4fc4c3eedd5d9a0..7eaf0280aff31f4f953d3f8e5bbe0570aed37eee 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -47,6 +47,9 @@ import java.util.logging.Level;
+ import java.util.logging.Logger;
+ import javax.imageio.ImageIO;
+ //import jline.console.ConsoleReader; // Paper
++import net.md_5.bungee.api.chat.ItemTag; // Paper
++import net.md_5.bungee.api.chat.TextComponent; // Paper
++import net.md_5.bungee.api.chat.hover.content.Content; // Paper
+ import net.minecraft.server.Advancement;
+ import net.minecraft.server.ArgumentEntity;
+ import net.minecraft.server.BiomeManager;
+@@ -88,6 +91,7 @@ import net.minecraft.server.MobSpawnerPatrol;
+ import net.minecraft.server.MobSpawnerPhantom;
+ import net.minecraft.server.MobSpawnerTrader;
+ import net.minecraft.server.NBTBase;
++import net.minecraft.server.NBTTagCompound; // Paper
+ import net.minecraft.server.PlayerList;
+ import net.minecraft.server.RegistryMaterials;
+ import net.minecraft.server.RegistryReadOps;
+@@ -2367,5 +2371,27 @@ public final class CraftServer implements Server {
+     public com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
+         return mobGoals;
+     }
++
++    @Override
++    public Content hoverContentOf(ItemStack itemStack) {
++        ItemTag itemTag = ItemTag.ofNbt(CraftItemStack.asNMSCopy(itemStack).save(new NBTTagCompound()).toString());
```

Comment:
```suggestion
+        ItemTag itemTag = ItemTag.ofNbt(CraftItemStack.asNMSCopy(itemStack).getTag().toString());
```

Reply:
No

Reply:
No that would not work.
That's what is called currently. By just getting the tag you would drop the type and the count.
```java
public NBTTagCompound save(NBTTagCompound nbttagcompound) {
        MinecraftKey minecraftkey = IRegistry.ITEM.getKey(this.getItem());

        nbttagcompound.setString("id", minecraftkey == null ? "minecraft:air" : minecraftkey.toString());
        nbttagcompound.setByte("Count", (byte) this.count);
        if (this.tag != null) {
            nbttagcompound.set("tag", this.tag.clone());
        }

        return nbttagcompound;
    }
```

Reply:
I just noticed `getTag()` returns null when the item doesn't have any additional data (e.g. enchantments, lores)
so, you have to check if the tag is null or not before `toString()`

Reply:
Thanks. Fixed by using `getOrCreateTag()` instead.